File: python/sort/mergesort.py
"""
Mergesort in Python
https://en.wikipedia.org/wiki/Merge_sort
"""


def merge(left, right):
    """
    Take two sorted lists and merge them into sorted list
    """
    left_len = len(left)
    right_len = len(right)
    min_len = min(left_len, right_len)
    merged_lis = []
    i, j = 0, 0
    while i < min_len and j < min_len:
        if left[i] < right[j]:
            merged_lis.append(left[i])
            i += 1
        else:
            merged_lis.append(right[j])
            j += 1
    merged_lis.extend(left[i:])
    merged_lis.extend(right[j:])
    return merged_lis


def mergesort(lis):
    """
    Take a list and return sorted list using mergesort. Doesn't modify original list
    """
    n = len(lis)
    if n == 1 or n == 0:  # If list is empty or single element is present then return list (Base case of recursion)
        return lis

    mid = n//2
    # Divide list in two parts: left_lis and right_lis at mid point
    left_lis = lis[:mid]
    right_lis = lis[mid:]
    left_lis = mergesort(left_lis)  # Recursively sort left_lis by mergesort
    right_lis = mergesort(right_lis)  # Recursively sort right_lis by mergesort
    mergesorted_lis = merge(left_lis, right_lis)  # Merge sorted lists
    return mergesorted_lis


def main():
    assert mergesort([4, 1, 2, 3, 9]) == [1, 2, 3, 4, 9]
    assert mergesort([1]) == [1]
    assert mergesort([2, 2, 1, -1, 0, 4, 5, 2]) == [-1, 0, 1, 2, 2, 2, 4, 5]


if __name__ == '__main__':
    main()

File: cpp/sort/merge_sort.cpp
/*
	Implementation of Merge Sort in C++(gcc 4.9.2)
	https://en.wikipedia.org/wiki/Merge_sort
*/

#include<iostream>
using namespace std;

int array[100000];

void merge(int l,int m,int r)
{
	int i,j,k,n1,n2;
	n1 = m-l+1;
	n2 = r-m;
	int L[n1],R[n2];
	for(i=0;i<n1;i++)
		L[i]=array[l+i];
	for(j=0;j<n2;j++)
		R[j]=array[m+j+1];
	i=0,j=0,k=l;
	while(i<n1 && j<n2)
	{
		if(L[i]<R[j])
		{
			array[k++]=L[i++];
		}
		else if(L[i]>R[j])
		{
			array[k++]=R[j++];
		}
		else
		{
			array[k++]=L[i++];
			array[k++]=R[i++];
		}
	}
	while(i<n1)
		array[k++]=L[i++];
	while(j<n2)
		array[k++]=R[j++];
}
void merge_sort(int l,int r)
{
	if(l<r)
	{
		int m = (l+r)/2;
		merge_sort(l,m);
		merge_sort(m+1,r);
		merge(l,m,r);
	}
}

int main()
{
	int n,i;
	cout<<"Input length of array- ";
	cin>>n;
	for(i=0;i<n;i++)
		cin>>array[i];
	merge_sort(0,n-1);
	cout<<"Sorted Array is-\n";
	for(i=0;i<n;i++)
		cout<<array[i]<<" ";
	return 0;
}